x/tokenomics/keeper/token_logic_modules.go

```diff
@@ -400,68 +442,88 @@ func (k Keeper) TokenLogicModuleGlobalMint(
 	); err != nil {
 		return nil, err
 	}
-	logger.Info(fmt.Sprintf("sent (%v) coins from the tokenomics module to the account with address %q", coinToAcc, addr))
+	logger.Info(fmt.Sprintf("sent (%v) coins from the tokenomics module to the account with address %q", coinToAcc, destAdr))
 
 	return &coinToAcc, nil
 }
 
-func (k Keeper) handleOverservicedApplication(
+// ensureClaimAmountLimits checks if the application was overserviced and handles
+// the case if it was.
+// Per Algorithm #1 in the Relay Mining paper, the maximum amount that a single supplier
+// can claim in a session is AppStake/NumSuppliersPerSession.
+// If this is not the case, then the supplier essentially did "free work" and the
+// actual claim amount is less than what was claimed.
+// Ref: https://arxiv.org/pdf/2305.10672
+func (k Keeper) ensureClaimAmountLimits(
 	ctx context.Context,
+	methodLogger log.Logger,
 	application *apptypes.Application,
-	settlementCoins cosmostypes.Coin,
+	supplier *sharedtypes.Supplier,
+	claimSettlementCoin cosmostypes.Coin,
 ) (
-	newSettlementCoins cosmostypes.Coin,
+	actualSettlementCoins cosmostypes.Coin,
 	err error,
 ) {
-	logger := k.Logger().With("method", "handleOverservicedApplication")
-	// over-serviced application
-	logger.Warn(fmt.Sprintf(
-		"THIS SHOULD NEVER HAPPEN. Application with address %s needs to be charged more than it has staked: %v > %v",
-		application.Address,
-		settlementCoins,
-		application.Stake,
-	))
-
-	// TODO_MAINNET(@Olshansk, @RawthiL): The application was over-serviced in the last session so it basically
-	// goes "into debt". Need to design a way to handle this when we implement
-	// probabilistic proofs and add all the parameter logic. Do we touch the application balance?
-	// Do we just let it go into debt? Do we penalize the application? Do we unstake it? Etc...
-	// See this document from @red-0ne and @bryanchriswhite for more context: notion.so/buildwithgrove/Off-chain-Application-Stake-Tracking-6a8bebb107db4f7f9dc62cbe7ba555f7
-	expectedBurn := settlementCoins
+	logger := methodLogger.With("helper", "ensureClaimAmountLimits")
+
+	// TODO_BETA_OR_MAINNET(@red-0ne): The application stake gets reduced with every claim
+	// settlement. Relay miners use the appStake at the beginning of a session to determine
+	// the maximum amount they can claim. We need to somehow access and propagate this
+	// value (via context?) so it is the same for all TLM processors for each claim.
+	// Note that this will also need to incorporate MintPerClaimGlobalInflation because
+	// applications are being overcharged by that amount in the meantime. Whatever the
+	// solution and implementation ends up being, make sure to KISS.
+	appStake := application.GetStake()
+
+	// Determine the max claimable amount for the supplier based on the application's stake
+	// in this session.
+	maxClaimableCoin := sdk.NewCoin(volatile.DenomuPOKT, appStake.Amount.Quo(math.NewInt(sessionkeeper.NumSupplierPerSession)))
```

Comment:
How do we ensure that applications/gateways know when they've reached the `maxClaimableCoin` for a given supplier such that they SHOULD/MUST send traffic to another supplier during a given session?

Is this solved and/or do we have an issue for it?

Reply:
```golang
// TODO_BETA(@red-0ne): When starting a new session, check what the MaxClaimableAmount
// (in uPOKT) by the Supplier as a function of
// (app_stake, compute_units_per_relay_for_service, global_compute_units_to_token_multiplier).
// TODO_CONFIG_NOTE: Whether or not the RelayMiner stop handling requests when the max is reached should be
// configurable by the operator.
// TODO_ERROR_NOTE: If overservicing is set to false, create a new error that the relay is rejected
// specifically because the supplier has reached the max claimable amount, so the caller should relay
// the request to another supplier.
```

See the `TODO_ERROR_NOTE`. No way to really enforce it but it's something!

Reply:
In `AppGateServer` this is not an issue due to the round-robin endpoint selection. `Path` needs to account for it.